src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs

```diff
@@ -95,12 +169,168 @@ public void Dispose()
+                        result = ReadTaskStartedEventArgs();
+                        break;
+                    case BinaryLogRecordKind.TaskFinished:
+                        result = ReadTaskFinishedEventArgs();
+                        break;
+                    case BinaryLogRecordKind.Error:
+                        result = ReadBuildErrorEventArgs();
+                        break;
+                    case BinaryLogRecordKind.Warning:
+                        result = ReadBuildWarningEventArgs();
+                        break;
+                    case BinaryLogRecordKind.Message:
+                        result = ReadBuildMessageEventArgs();
+                        break;
+                    case BinaryLogRecordKind.CriticalBuildMessage:
+                        result = ReadCriticalBuildMessageEventArgs();
+                        break;
+                    case BinaryLogRecordKind.TaskCommandLine:
+                        result = ReadTaskCommandLineEventArgs();
+                        break;
+                    case BinaryLogRecordKind.TaskParameter:
+                        result = ReadTaskParameterEventArgs();
+                        break;
+                    case BinaryLogRecordKind.ProjectEvaluationStarted:
+                        result = ReadProjectEvaluationStartedEventArgs();
+                        break;
+                    case BinaryLogRecordKind.ProjectEvaluationFinished:
+                        result = ReadProjectEvaluationFinishedEventArgs();
+                        break;
+                    case BinaryLogRecordKind.ProjectImported:
+                        result = ReadProjectImportedEventArgs();
+                        break;
+                    case BinaryLogRecordKind.TargetSkipped:
+                        result = ReadTargetSkippedEventArgs();
+                        break;
+                    case BinaryLogRecordKind.EnvironmentVariableRead:
+                        result = ReadEnvironmentVariableReadEventArgs();
+                        break;
+                    case BinaryLogRecordKind.ResponseFileUsed:
+                        result = ReadResponseFileUsedEventArgs();
+                        break;
+                    case BinaryLogRecordKind.PropertyReassignment:
+                        result = ReadPropertyReassignmentEventArgs();
+                        break;
+                    case BinaryLogRecordKind.UninitializedPropertyRead:
+                        result = ReadUninitializedPropertyReadEventArgs();
+                        break;
+                    case BinaryLogRecordKind.PropertyInitialValueSet:
+                        result = ReadPropertyInitialValueSetEventArgs();
+                        break;
+                    case BinaryLogRecordKind.AssemblyLoad:
+                        result = ReadAssemblyLoadEventArgs();
+                        break;
+                    default:
+                        OnRecoverableReadError?.Invoke(
```

Comment:
It is little bit inconsistent. Here, we raise `OnRecoverableReadErro` even if it is not recoverable while bellow we only throw exception and not raise it. I think we can consider to raise error also before we throw that InvalidDataException

Reply:
I'll try to be brave and will go the other way - the `OnRecoverableReadError` is only raised when we can continue. Exception is thrown otherwise. Allways. That means we break the scenario where unknown events lead to simple end of reading (returning of null) - which is dangerous and I believe incorrect behavior.
It might break few of our tests (then will then need to set the forwardCompatibility flag) that uses this for some reason. But other than that it should not break sane scenario.

Thoughts?